### PR TITLE
Juju 2 support

### PR DESCRIPTION
**Pull request by arturo-seijas**
Applicable spec: N/A

### Overview

Support for Juju 2

### Rationale

N/A

### Juju Events Changes

N/A

### Module Changes

charm.py

### Library Changes

N/A

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -97,6 +97,17 @@ deps =
 commands =
     bandit -c {toxinidir}/pyproject.toml -r {[vars]src_path} {[vars]tst_path}
 
+[testenv:integration-juju-2]
+description = Run Juju 2 integration tests
+deps =
+    juju==3.2.2
```
[the remark this answers is not in the document]
Comment (arturo-seijas) [reply tox.ini:103]: Good point

File: tests/unit/test_charm.py [hunk 2]
```diff
@@ -120,12 +122,14 @@ def test_legacy_relation_joined_when_leader():
     assert "password_id" not in data
 
 
-def test_relation_joined_when_leader():
+@patch.object(ops.JujuVersion, "from_environ")
```
[the remark this answers is not in the document]
Comment (arturo-seijas) [reply tests/unit/test_charm.py:125]: What's the difference/advantages?

[the remark this answers is not in the document]
Comment (weiiwang01) [reply tests/unit/test_charm.py:125]: Not much advantage other than it's one line shorter, up to you :)

**Review by weiiwang01 — DISMISSED** [review]

File: tests/unit/test_charm.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (arturo-seijas) [reply tests/unit/test_charm.py:125]: I'm all for saving some lines, but on the other hand, this narrows down where an error would appear

File: .github/workflows/integration_test.yaml [hunk 3]
```diff
@@ -8,11 +8,16 @@ jobs:
...
       juju-channel: 3.1/stable
+  integration-tests-juju2:
+    name: Integration test with juju 2.9
+    uses: canonical/operator-workflows/.github/workflows/integration_test.yaml@main
+    secrets: inherit
+    with:
+      channel: 1.28/stable
+      juju-channel: 2.9/stable
+      pre-run-script: scripts/pre-integration-test.sh
+      test-tox-env: integration-juju-2
```
Comment (gregory-schiano) [inline .github/workflows/integration_test.yaml:23]: Can't we have a "matrix" based approached here ?

Comment (arturo-seijas) [reply .github/workflows/integration_test.yaml:23]: I'm afraid we can't, we need a specific version of juju with an specific version of k8s, not the cartesian product with all combinations

Comment (gregory-schiano) [reply .github/workflows/integration_test.yaml:23]: Well there are ways to do that but probably not worthwhile for 2 cases

**Review by gregory-schiano — DISMISSED** [review]
LGTM

**Review by gregory-schiano — DISMISSED** [review]
LGTM

**Review by weiiwang01 — DISMISSED** [review]

**Review by weiiwang01 — APPROVED** [review]

**Review by gregory-schiano — APPROVED** [review]
LGTM

**Review by cbartz — APPROVED** [review]
LGTM

**Pull request merged**